sway-core/src/semantic_analysis/ast_node/expression/typed_expression.rs

```diff
@@ -2673,41 +2712,56 @@ impl ty::TyExpression {
+                    if let TypeInfo::Ref {
+                        referenced_type, ..
+                    } = actually
                     {
-                        // SAFETY: safe by the guard above
-                        let array_length = array_length
-                            .as_literal_val()
-                            .expect("unexpected non literal array length")
-                            as u64;
-                        if index_literal >= array_length {
-                            return Err(handler.emit_err(CompileError::ArrayOutOfBounds {
-                                index: index_literal,
-                                count: array_length,
-                                span: index.span.clone(),
+                        actually = (*engines.te().get(referenced_type.type_id)).clone();
+                    }
+                    match actually {
+                        TypeInfo::Array(elem_ty, array_length)
+                            if array_length.as_literal_val().is_some() =>
+                        {
+                            parent_rover = symbol;
+                            symbol = elem_ty.type_id;
+                            symbol_span = index_span.clone();
+
+                            if let Some(index_literal) = index
+                                .expression
+                                .as_literal()
+                                .and_then(|x| x.cast_value_to_u64())
+                            {
+                                // SAFETY: safe by the guard above
+                                let array_length = array_length
+                                    .as_literal_val()
+                                    .expect("unexpected non literal array length")
+                                    as u64;
+                                if index_literal >= array_length {
+                                    return Err(handler.emit_err(CompileError::ArrayOutOfBounds {
+                                        index: index_literal,
+                                        count: array_length,
+                                        span: index.span.clone(),
+                                    }));
+                                }
+                            }
+
+                            // `index_span` does not contain the enclosing square brackets.
+                            // Which means, if this array index access is the last one before the
+                            // erroneous expression, the `full_span_for_error` will be missing the
+                            // closing `]`. We can live with this small glitch so far. To fix it,
+                            // we would need to bring the full span of the index all the way from
+                            // the parsing stage. An effort that doesn't pay off at the moment.
+                            // TODO: Include the closing square bracket into the error span.
```

Comment:
Do we have an issue for this TODO? If not, we should create one just to keep track of it. In any case, we should add a link to the issue in the comment, so that we can track it that way.